### PR TITLE
Support beacon proxies for Truffle

### PR DESCRIPTION
**Pull request by ericglau**
Fixes https://github.com/OpenZeppelin/openzeppelin-upgrades/issues/342

Adds the following APIs to the Truffle plugin (see [doc](https://github.com/ericglau/openzeppelin-upgrades/blob/truffle/docs/modules/ROOT/pages/api-truffle-upgrades.adoc)):
`deployBeacon`
`upgradeBeacon`
`deployBeaconProxy`
`loadProxy` (supports UUPS/Transparent/Beacon proxies)

Modifies the following API:
`prepareUpgrade` (accepts a UUPS/Transparent/Beacon proxy address or a Beacon address)

Adds the following API for Truffle and Hardhat:
`beacon.getImplementationAddress`

Refactors some common functions into core.

### CODE REVIEW

**Comment by ericglau** [comment]
The new test cases need to be updated to more properly reflect Truffle migration patterns.

**Comment by ericglau** [comment]
Ready for review

**Review by frangio — COMMENTED** [review]
Great work! Minimal comments.

Let's follow up with a separate PR to remove the interface stuff as discussed earlier.

I did not look at the tests, let me know if you want me to review anything in particular in them.

File: packages/plugin-truffle/src/deploy-beacon.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+export async function deployBeacon(Contract: ContractClass, opts: Options = {}): Promise<ContractInstance> {
+  const { impl } = await deployBeaconImpl(Contract, opts);
+
+  const { deployer } = withDefaults(opts);
+  const UpgradeableBeaconFactory = getUpgradeableBeaconFactory(Contract);
+  const beaconDeployment = await deploy(deployer, UpgradeableBeaconFactory, impl);
+
+  Contract.address = beaconDeployment.address;
+  const beaconContract = new Contract(beaconDeployment.address);
```
[the remark this answers is not in the document]
Comment (ericglau) [reply packages/plugin-truffle/src/deploy-beacon.ts:19]: Calling `new Contract(address)` creates a `ContractInstance` (according to [this](https://github.com/OpenZeppelin/openzeppelin-upgrades/blob/master/packages/plugin-truffle/src/utils/truffle.ts#L9)).  Is that not what we should be returning?

This follows a similar pattern to what was done [here](https://github.com/OpenZeppelin/openzeppelin-upgrades/blob/b8dfb644a7b83f7bbeafb20f7f4762940f495874/packages/plugin-truffle/src/deploy-proxy.ts#L69) previously for deployProxy().

[the remark this answers is not in the document]
Comment (frangio) [reply packages/plugin-truffle/src/deploy-beacon.ts:19]: The issue is that a beacon is not a proxy. A beacon will not respond to the interface of the "Contract" here, so it's not right to wrap the beacon address in a Contract instance.

[the remark this answers is not in the document]
Comment (ericglau) [reply packages/plugin-truffle/src/deploy-beacon.ts]: Fixed

File: packages/plugin-truffle/src/deploy-beacon.ts [hunk 2]
```diff
@@ -15,8 +15,8 @@ export async function deployBeacon(Contract: ContractClass, opts: Options = {}):
   const UpgradeableBeaconFactory = getUpgradeableBeaconFactory(Contract);
   const beaconDeployment = await deploy(deployer, UpgradeableBeaconFactory, impl);
 
-  Contract.address = beaconDeployment.address;
-  const beaconContract = new Contract(beaconDeployment.address);
+  UpgradeableBeaconFactory.address = beaconDeployment.address;
```
Comment (frangio) [inline packages/plugin-truffle/src/deploy-beacon.ts:18]: As far as I know this line won't have any effect. `deployed()` is only tracked for the user's contracts but the artifact for this is deep inside our code, not in the user's artifacts directory.

**Review by frangio — APPROVED** [review]
The line I mentioned could be removed I think but this is good!

**Pull request merged**